File: Warmup/fortran/arctan.f90
program arctan
! Program to compute the first derivative of arctan(x) at x=sqrt(2)

IMPLICIT NONE

  integer :: i,j,h_steps
  real(8) :: x,darctan !darctan is the exact first derivative
  real(4) :: h_single
  real(4),dimension(:,:),allocatable :: single
  !single(:,1) is for h, (:,2) is for the derivative
  real(8) :: h_double
  real(8),dimension(:,:),allocatable :: double
  !double(:,1) is for h, (:,2) is for the derivative
  
  x = 0.0
  h_steps = 0
  print *, "Find first derivative of arctan(x)"
  !read in x
  write (*,'(A5)',advance="no") "x = "
  read *, x
  !read in h_steps
  print *, "h starts from 0.1"
  write (*,'(A19)',advance="no") "Number of steps = "
  read *, h_steps

  !initialize h
  h_single = 0.1
  h_double = 0.1d0
  
  allocate(single(h_steps,2),double(h_steps,2))
  
  do i=1, h_steps
    !single precision
    single(i,1) = h_single
!    single(i,2) = (atan(x + h_single)-atan(x))/h_single
    single(i,2) = (atan(x + h_single)-atan(x - h_single))/(2*h_single)
    h_single = h_single * 0.5
    !double precision
    double(i,1) = h_double
!    double(i,2) = (atan(x + h_double)-atan(x))/h_double
    double(i,2) = (atan(x + h_double)-atan(x - h_double))/(2*h_double)
    h_double = h_double * 0.5
  end do

  open(unit=11, file="out-single_3.dat", status="replace" )
  open(unit=12, file="out-double_3.dat", status="replace" )
  !get the exact first derivative
  darctan = 1/(1+x*x)
  do i=1, h_steps
    write(11,'(E16.10,2X,E16.10)') log10(single(i,1)), &
      & log10(abs((single(i,2)-darctan)/(darctan)))
    write(12,'(E16.10,2X,E16.10)') log10(double(i,1)), &
      & log10(abs((double(i,2)-darctan)/(darctan)))
  end do
  close(11)
  close(12)

  deallocate(single, double)

end program arctan

File: Warmup/arctan.cpp
//Program to compute first derivative of arctan(x) at $x = \sqrt(2)$ with two formulae
using namespace std;

#include <iostream>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <stdio.h>

ofstream ofile;

// function to get derivative
void derivative(double x,int num,double initial_step,float *h_single,float *f2_single,double *h_double,double *f2_double)
{
	int loop;
	double h = initial_step;
	for (loop=0; loop < num; loop++) {
		h_single[loop] = h;
		f2_single[loop] = (atan(x+h) - atan(x-h))/(2*h);
		h_double[loop] = h;
		f2_double[loop] = (atan(x+h) - atan(x-h))/(2*h);
		h = h*0.5;
	}
}

// function for output
void output(double x,int num,float *h_single,float *f2_single,double *h_double,double *f2_double)
{
	int i;
	double epsilon = 1/(1+x*x);
	ofile << setiosflags(ios::showpoint | ios::uppercase);
	//for relative errors
	float err_single;
	double err_double;
	for (i=0; i<num; i++){
		err_single = log10(fabs((f2_single[i]-epsilon)/epsilon));
		ofile << setw(15) << setprecision(8) << log10(h_single[i]);
		ofile << setw(15) << setprecision(8) << err_single;
		err_double = log10(fabs((f2_double[i]-epsilon)/epsilon));
		ofile << setw(15) << setprecision(8) << log10(h_double[i]);
		ofile << setw(15) << setprecision(8) << err_double << endl;
	}
	
}


// ********** main program ***********
int main()
{
	int num; //number of steps 
	double x;
	const double initial_step = 0.1; // h_initial=0.1
	float *h_single, *f2_single;
	double *h_double, *f2_double;

	//Get the initial information
	cout << "x = " ;
	cin >> x;
	cout << "number of steps: ";
	cin >> num;

	//get the vector of h and corresponding derivatives
	h_single = new float[num];
	f2_single = new float[num];
	h_double = new double[num];
	f2_double = new double[num];
	
	//call the function to get the derivative
	derivative(x,num,initial_step,h_single,f2_single,h_double,f2_double);
	
	//output
	ofile.open("out-threepoint.dat");
	
	output(x,num,h_single,f2_single,h_double,f2_double);

	ofile.close();

	delete [] h_single;
	delete [] h_double;
	delete [] f2_single;
	delete [] f2_double;

	return 0;
} //end of main program
	
